Show user details

weedySeaDragon opened the pull request:
PT Story: https://www.pivotaltracker.com/story/show/140358959


Changes proposed in this pull request:
1.  add `show` action to the UsersController and the associated route
2. moved entries for User attributes to the `activerecord.attributes.users` section in the locale files;  referred `users.index` and `users.show` to them as much as possible
   - **NEEDS REVIEWING** 
3. added a feature for what should be shown
4. added a helper for showing the most recent login info (and spec for it)
5. added a view to show the details about a user
6. added a stylesheet so if a user is an admin, their email is in RED
7. on the 'all users' (`index`) page, linked the email for each user to the detailed page


## Screenshots:
**Admin (notice the email is styled with CSS [red, bold]):**
<img width="609" alt="sv-admin-neverreset" src="https://cloud.githubusercontent.com/assets/673794/23240155/8defa0e4-f920-11e6-8593-262cabca53ab.png">

<img width="607" alt="en-admin-neverreset" src="https://cloud.githubusercontent.com/assets/673794/23240156/8df020c8-f920-11e6-9fc1-a09edf370e17.png">


**A user that has never logged in:**

<img width="611" alt="sv-user-never-never" src="https://cloud.githubusercontent.com/assets/673794/23240154/8def13f4-f920-11e6-83a2-de4e3b458b42.png">

<img width="615" alt="en-user-never-never" src="https://cloud.githubusercontent.com/assets/673794/23240153/8dd3a3c6-f920-11e6-9d24-8bcba9ca09f5.png">


**A user that *has* logged in and has had the password reset:**

<img width="650" alt="sv-user-loggedin-reset" src="https://cloud.githubusercontent.com/assets/673794/23240237/366b32e2-f921-11e6-88a3-3adb81b7f83d.png">

<img width="677" alt="en-user-loggedin-reset" src="https://cloud.githubusercontent.com/assets/673794/23240238/366c7d46-f921-11e6-8969-bd06119b0e96.png">





Ready for review:
@patmbolger

Commit:
merge from AgileVentures/develop

Commit:
merge from AgileVentures/develop

weedySeaDragon committed:
Business category form i18 name and description (#145)

* put 'Kategory/Business Category' in front of 'name' and 'description'; no separate form entries needed

* refer to business_category: attribute entries (instead of separate ... form entries)

* just show 'Namn'/'Name' and 'Beskrivning'/'Description' for the category attributes (no need to start each with 'Kategori ...'

weedySeaDragon committed:
Uploaded File: file size is validated (can't be larger than 5 MB) #151)

* testing for size validation

(cherry picked from commit 5e4c622)

* add fixture: file just under the upload size limit

(cherry picked from commit 146f88e)

* step for specifying max_size with the validation error t() message

(cherry picked from commit c585fce)

* file size validation error messages

* show :file_too_large error for size validation error

* clean up steps for specifying locale; remove hardcoded 'sv' locale step

* scenarios for different locales (error message in diff. language), upload into existing application as well as new application

* don't show 'success' for new/update if there was a problem with the file upload

* improve code (better checking for nil)

* added explanation of the size validation error messages and the variables available for them

weedySeaDragon committed:
Only show 'complete' companies (bugfix) (#153)

* add scope for complete and a method that will check of a company is complete

* #index: only list companies that are complete (must use SQL so that pagination remains correct)

* only display (and try to sort) companies that are complete; they will have a name that is not nil

* assume @companies has been scoped as needed (ex: maybe by using .complete)

* name could be nil or empty, old_region could be nil or empty

* add tests for .complete scope and #complete? method

* only sort and show complete companies; only show region name if there is a region (not nil)

* steps to set region to nil, old_region to nil or a particular string

* do not list cities or names of 'incomplete' companies

* + Feature: only show complete companies

* add needed space in .haml (and make HoundCI happy)

* controller should set @companies to only complete companies, and sort them (move out of view)

weedySeaDragon committed:
Only show org nr to admin (#158)

* merge from AgileVentures/develop

* add "I should not see" the company number (unless the admin)

* a feature for 'show company details'

* only show the org. nr. if user is the admin

weedySeaDragon committed:
Dev Doc: Development and Production servers (#140)

* very first draft of documenting development and production servers for the project

* very first draft of documenting development and production servers for the project

* use 'fetches'; fix typo

* removed 'TODO' so Code climate won't complain

Commit:
Merge remote-tracking branch 'AgileVentures/develop' into origin-develop

Commit:
Merge remote-tracking branch 'AgileVentures/develop' into origin-develop

Commit:
add steps: 'user details' for a page for a user (email); when a user is or was signed in, and how many times

Commit:
feature:  Show User Details

Commit:
add :show route for users resource

Commit:
controller: add before_actions to set a specific @user and to authorize the current user

Commit:
move terms to the attributes for a User; have users.index and user.show refer to them as much as possible

Commit:
move terms to the attributes for a User; have users.index and user.show refer to them as much as possible

Commit:
link the email for a user to the detail page for that user

Commit:
show detals for a user

Commit:
CSS style for is-admin (and precompile this asset)

Commit:
add 'users' stylesheet

Commit:
if the user is an admin, use CSS to style the email so it's obvious

Commit:
remove ?

Commit:
remove ?

Commit:
show the most recent login time and IP

Commit:
just show the most recent time and IP

Commit:
use Hex representation of color (per Hundbot)

Commit:
remove excess blank lines (but not too many!)

weedySeaDragon commented on app/views/users/show.html.haml:
```diff
@@ -0,0 +1,44 @@
+      = @user.sign_in_count
+
+  - if @user.admin?
+    %p.is-admin #{t('.is_an_admin')}
```
I wasn't able to make it work like that, Houndbot.

patmbolger replied:
I was able to get this to work with:
```
%p.is-admin= t('.is_an_admin')
```

weedySeaDragon replied:
Thanks.  That's helpful. 

I keep struggling with this kind of thing in HAML.  (@thesuss has helped me with this same syntax issue many times.)  Time for me to go read thru the HAML specs and practice with some tutorials. :-)

Commit:
using variable to refer to a hex color (per houndbot)

patmbolger commented on app/views/users/index.html.haml:
```diff
@@ -15,7 +15,7 @@
-          %td= user.email
+          %td= link_to user.email, user
           %td
             %small= #{time_ago_in_words(user.created_at)} ago
```
Please change this line to:
```
%small= "#{time_ago_in_words(user.created_at)} ago"
```
(add double quotes around the string - not sure why this stopped working (nothing to do with your PR)).

patmbolger commented on app/views/users/show.html.haml:
```diff
@@ -0,0 +1,44 @@
+      (#{time_ago_in_words(most_recent_login_time @user)} ago)
+    %p
+      %b #{t('.last_login_ip')}:
+      #{ most_recent_login_ip @user}
```
Showing login IP might be overkill - we should confirm with client.

weedySeaDragon replied:
Agreed.  

@thesuss  -- thoughts on this?  Has there ever been a situation when it was helpful to see the IP address, or might be?  

I'd be happy to remove this for now and then also make sure we explicitly ask the client about it.   Seems like a pretty small thing and would be easy to add back in.  (would be a good 'easy change' for a novice.)

I'll make a chore in PivotalTracker so that we remember to ask this in a client meeting.  Seems low priority, so I don't see a need to add it into the upcoming, already full, meeting. Happy to do something different, though.

patmbolger commented on app/views/users/show.html.haml:
```diff
@@ -0,0 +1,44 @@
+      = @user.sign_in_count
+
+  - if @user.admin?
+    %p.is-admin #{t('.is_an_admin')}
```
Do we need this since it says we are admin at top of page?  (and the email at top is also shown in red)

weedySeaDragon replied:
_(I initially misunderstood your comment.)_

I moved the info up next to the title (email).  See the updated screenshot

patmbolger commented on app/controllers/users_controller.rb:
```diff
@@ -1,7 +1,24 @@
+
+
   def index
     authorize User
```
Can delete this line since we're now using the `before_action` hook.

patmbolger commented on app/helpers/users_helper.rb:
```diff
@@ -0,0 +1,11 @@
+
+  def most_recent_login_time user
+    user.current_sign_in_at.blank? ? user.last_sign_in_at : user.current_sign_in_at
+  end
```
We should use this helper in the index view as well. (separate story?)

weedySeaDragon replied:
I'd like this to be a separate story just to keep things simple.

weedySeaDragon replied:
I've created a PT chore for this one:  https://www.pivotaltracker.com/story/show/140897375

patmbolger commented on spec/helpers/users_helper_spec.rb:
```diff
@@ -0,0 +1,63 @@
+RSpec.describe UsersHelper, type: :helper do
+  let!(:user) { create(:user) }
+  let(:right_now) { Time.now }
+  let(:yesterday) { Time.now - 60*60*24 - 120}
```
You can do this with:
```
Time.now - 1.day - 2.minutes
```
(easier to read)

patmbolger reviewed (changes requested):
Please see comments

Commit:
Merge branch 'develop' into origin-develop

Commit:
Merge branch 'origin-develop' into sprint8-140358959-show-user-details

Commit:
add helper to show the 'time ago in words' localized correctly

Commit:
remove helper for showing IP address from most recent login (not used)

Commit:
spec for showing 'time ago in words' localized

Commit:
add entries:  time_ago,  password_never_reset

Commit:
add scenario for if the password was/wasn't reset; add step to set data; remove assertions for showing the IP address

Commit:
remove IP address; note is admin next to title; i18n 'time ago'; show words if password was never reset

Commit:
use words to for parts of time instead of stuff like "60*60*24"

thesuss commented on config/locales/sv.yml:
```diff
@@ -384,14 +395,31 @@ sv:
       created: *created
-      last_login: Senast inloggad
-      logged_in_count: Inloggningar
+      last_login: *user_last_login_ip
```
should not be _ip here, right?

Commit:
fix users.index.last_login   (c&p error)

Commit:
authorization is called before all actions; no need to explictly also call it in #index

weedySeaDragon commented:
@patmbolger or @thesuss  -- Is this OK now?  If so, can it be merged please? 
 It's blocking me from doing **"Admin can change password for user"** https://www.pivotaltracker.com/story/show/139158857

patmbolger commented on app/helpers/application_helper.rb:
```diff
@@ -32,4 +32,9 @@ def assocation_empty?(assoc)
   end
+
+  def i18n_time_ago_in_words(past_time)
+    "#{time_ago_in_words(past_time)} #{t('time_ago')}"
```
Is this correct wording in Swedish (Susanna)?

If this works we're OK.  Generally, think it's better to pass in the `time_ago_in_words(past_time)` as a parameter to `t('time_ago')` in order to prevent "lego translation". (https://www.mediawiki.org/wiki/Localisation#Avoid_fragmented_or_.27patchwork.27_messages)

thesuss replied:
Sorry for the delay. I have no idea, what is/are the wording/s?

thesuss replied:
If not we can fix it in a separate PR? Or do as is generally better in a separate PR?

weedySeaDragon replied:
Good idea, thesuss: I'll fix this, then we can do a separate PR for the the translations.  (We'll be able to keep going and it'll hopefully be easy for you to just focus on the translation stuff when it's the only stuff in the PR.)

patmbolger replied:
@thesuss - what I meant was, sometimes if translations (or word sequences) are stuck together (like sticking lego toy pieces together) you can get mangled wording for a particular language.

For instance, here the wording for English would be `3 days ago`.

However, lets say that the equivalent, correct phrase in Swedish would be `past time 3 days` (totally made up, of course), but with lego construction it would be `3 days past time` (and hence incorrect).

In any case, if the correct approach works fine in Swedish then we don't have a problem - especially since we don't anticipate supporting other languages.

patmbolger reviewed (commented):
I think this is good to go - just have one question for Susanna regarding translation helper.

weedySeaDragon committed:
pass amount_of_time to time_ago so the exact grammar/order can be adjusted for a locale as needed

weedySeaDragon committed:
Merge branch 'develop' into sprint8-140358959-show-user-details

weedySeaDragon commented:
@patmbolger Thanks for the excellent suggestion and link about the translation.  I've updated the code (please check the commit) and also added the link into [the wiki page about i18n/l10n](https://github.com/AgileVentures/shf-project/wiki/How-to-Use-I18n-to-Localize-SHF).

I made a separate PT chore for Susanna to check the translation so that we can merge this and she can get to the translation when she's able.https://www.pivotaltracker.com/story/show/140961225

weedySeaDragon committed:
Merge branch 'develop' into sprint8-140358959-show-user-details

thesuss commented on config/locales/sv.yml:
```diff
@@ -79,6 +82,14 @@ sv:
         email: &email E-post
         password: *password
         created: &created Skapad
+        current_sign_in_at: &user_current_signed_in_time Inloggad pä
```
I did find some issues :) 
I think pä is supposed to be på. But ut can be omitted. 

So, line 85: "Inloggad"

thesuss commented on config/locales/sv.yml:
```diff
@@ -79,6 +82,14 @@ sv:
         password: *password
         created: &created Skapad
+        current_sign_in_at: &user_current_signed_in_time Inloggad pä
+        current_sign_in_ip: &user_current_ip Inloggad pä IP
```
Inloggad med IP

thesuss commented on config/locales/sv.yml:
```diff
@@ -79,6 +82,14 @@ sv:
+        current_sign_in_at: &user_current_signed_in_time Inloggad pä
+        current_sign_in_ip: &user_current_ip Inloggad pä IP
+        last_login: &user_last_login Senast inloggad
+        last_login_ip: &user_last_login_ip Senast inloggad IP
```
Senast inloggad med IP

thesuss commented on config/locales/sv.yml:
```diff
@@ -79,6 +82,14 @@ sv:
+        last_login_ip: &user_last_login_ip Senast inloggad IP
+        sign_in_count: &sign_in_count Inloggningar
+        created: *created
+        reset_password_sent_at: &user_reset_password_sent_at Återställning av lösenord information sändes
```
No biggie, but while we're at it: 
Information angående återställning av lösenord sändes

patmbolger commented:
This looks good to me.  Looks like Susanna has some outstanding translation changes.  

Also, looks like it needs a rebase due to conflicts with upstream/develop

weedySeaDragon committed:
sv corrections

weedySeaDragon committed:
Merge branch 'develop' into sprint8-140358959-show-user-details

weedySeaDragon committed:
resolve merge conflicts

weedySeaDragon committed:
Merge branch 'develop' into sprint8-140358959-show-user-details
